README
Tk/GraphViz version 0.01
========================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires the 'dot' program that is a part.  It may also be used in conjuction with the GraphViz module by Leon Brocard.

Being a Tk widget, Tk is required as well.


COPYRIGHT AND LICENCE

Copyright (C) 2003 Jeremy Slade

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
- add prerender hook

1.09 2020-12-24
- nodes method
- edges method
- promote JGS's useful t/compareGraphs to scripts/tkgraphviz
- use Tk tags correctly

1.08 2020-12-23
- scrollTo method

1.07 2020-11-17
- fix edge-parsing on Really Big coordinates (with e-notation)

1.06 2020-11-16
- fix accidental double-run of dot with async=0

1.05 2020-11-16
- fix handling of \-quoting in labels
- record label justification now correct
- records have fillcolor, font, etc respected
- use Tk::IO, not own complicated scheme
- createBindings returns $self for chaining
- ->show no longer automatically calls ->fit by default
- implement Msquare and Mdiamond

1.04 2020-11-08
- tests only run when can do display stuff, avoiding false negatives
- no install if no DOT

1.03 2020-11-08
- support a GraphViz2 object as input

1.02 2020-11-07
- parsing improvements to account for DOT 2.43 changing output format a bit
- parsing fix from RT#79377 - thanks @geoffhart
- ->show also calls MainWindow->update just in case - RT#64089

1.01  Mon Dec  8 21:32:21 MST 2008
	- Added 'folder' and 'component' node shapes (Jogi Kuenstner)

1.00  Sat Mar 29 12:38:22 MDT 2008
	- Use layout_cmd as an array so shell isn't require to
	  exec it (Reiner Post);
	- Fixed used of fillcolor vs color for filled nodes (Phil Kasten);
	- Added missing 'invtrapezium' shape (Charles Minc)

0.90  Thu Apr 09 23:27:55 MDT 2007
	- Fixes for zooming
	- Changes to support fontcolor, arrowhead/arrowtail (Tobias Lorenz)
	
0.15  Thu Jun 23 11:12:20 MDT 2005
	- Fixed issue with stopping dot subprocess already in progress
	
0.14  Thu Apr 21 14:12:58 MDT 2005
	- Use '>&STDERR' with IPC::Open3::open3()
	
0.13  Tue Feb 15 22:24:15 MST 2005
	- Handle negative font sizes (John Cerney)
	- Better DOS/Win support (John Cerney)

0.11  Thu Sep 16 15:46:18 PDT 2004
	- use reaper to handle child processes

0.10  Wed Feb  5 10:06:00 2003
	- asynchronous layout mode

0.07  Thu Sep 18 09:53:16 2003
	- Minor clean-up of some warnings and such
	- Doc updates

0.06  Wed Mar  5 23:54:55 2003
	- Supports more shape types, still not all of them though
	- Added options to show() to allow use of neato for layout

0.05  Tue Mar  4 21:27:39 2003
	- Added -keypad bindings option
        - Implemented record nodes, font scaling, bezier curves (John Cerney)

0.04  Tue Feb 25 21:46:45 2003
	- A few minor enhancements to existing functionality

0.03  Wed Feb  5 10:06:00 2003
	- Original version packaged for submission to CPAN

0.01  Sat Jan 25 14:28:47 2003
	- Original version packaged for submission to CPAN



Makefile.PL
use strict;
use warnings;
use File::Spec;
use File::Temp;
use ExtUtils::MakeMaker;

sub is_dot_installed {
  # 1: Create a temp file containing DOT commands.
  # The EXLOCK option is for BSD-based systems.
  # newdir() croaks() if it fails, which is what we want.
  my $temp_dir = File::Temp->newdir('temp.XXXX', CLEANUP => 1, EXLOCK => 0, TMPDIR => 1);
  my $gv_file = File::Spec->catfile($temp_dir, 'test.gv');
  open my $fh, ">", $gv_file or die "Can't create temp file: $!\n";
  print $fh "digraph graph_14 {node_14}\n" or die "Can't write to temp file: $!\n";
  close $fh or die "Can't close temp file: $!\n";
  my $stdout = `dot -Tsvg $gv_file`;
  die "Please install Graphviz from http://www.graphviz.org/\n" if $stdout !~ m|</svg>|;
}

is_dot_installed();

my $repo = 'graphviz-perl/Tk-GraphViz';
WriteMakefile(
  NAME => 'Tk::GraphViz',
  VERSION_FROM => 'lib/Tk/GraphViz.pm',
  PREREQ_PM => {
    Tk => '800.020',
  },
  ABSTRACT => 'Tk widget (derived from Canvas) for displaying graphs in the GraphViz format.',
  AUTHOR => 'Jeremy Slade <jeremy@jkslade.net>',
  EXE_FILES => [ glob 'scripts/*' ],
  META_MERGE => {
    "meta-spec" => { version => 2 },
    dynamic_config => 0,
    resources => {
      x_IRC => 'irc://irc.perl.org/#graphql-perl',
      repository => {
        type => 'git',
        url => "git\@github.com:$repo.git",
        web => "https://github.com/$repo",
      },
      bugtracker  => {
        web => "https://github.com/$repo/issues",
      },
      license => [ 'http://dev.perl.org/licenses/' ],
    },
    prereqs => {
      develop => {
        requires => {
          'Test::Pod::Coverage' => '1.08',
          'Test::Pod' => '1.22',
        },
      },
    },
  },
);


lib/Tk/GraphViz/parseRecordLabel.pm
####################################################################
#
#    This file was generated using Parse::Yapp version 1.21.
#
#        Don't edit this file, use source file instead.
#
#             ANY CHANGE MADE HERE WILL BE LOST !
#
####################################################################
package Tk::GraphViz::parseRecordLabel;
use vars qw ( @ISA );
use strict;

@ISA= qw ( Parse::Yapp::Driver );
#Included Parse/Yapp/Driver.pm file----------------------------------------
{
#
# Module Parse::Yapp::Driver
#
# This module is part of the Parse::Yapp package available on your
# nearest CPAN
#
# Any use of this module in a standalone parser make the included
# text under the same copyright as the Parse::Yapp module itself.
#
# This notice should remain unchanged.
#
# Copyright © 1998, 1999, 2000, 2001, Francois Desarmenien.
# Copyright © 2017 William N. Braswell, Jr.
# (see the pod text in Parse::Yapp module for use and distribution rights)
#

package # hide from PAUSE
  Parse::Yapp::Driver;

require 5.004;

use strict;

use vars qw ( $VERSION $COMPATIBLE $FILENAME );

# CORRELATION #py001: $VERSION must be changed in both Parse::Yapp & Parse::Yapp::Driver
$VERSION = '1.21';
$COMPATIBLE = '0.07';
$FILENAME=__FILE__;

use Carp;

#Known parameters, all starting with YY (leading YY will be discarded)
my(%params)=(YYLEX => 'CODE', 'YYERROR' => 'CODE', YYVERSION => '',
			 YYRULES => 'ARRAY', YYSTATES => 'ARRAY', YYDEBUG => '');
#Mandatory parameters
my(@params)=('LEX','RULES','STATES');

sub new {
    my($class)=shift;
	my($errst,$nberr,$token,$value,$check,$dotpos);
    my($self)={ ERROR => \&_Error,
				ERRST => \$errst,
                NBERR => \$nberr,
				TOKEN => \$token,
				VALUE => \$value,
				DOTPOS => \$dotpos,
				STACK => [],
				DEBUG => 0,
				CHECK => \$check };

	_CheckParams( [], \%params, \@_, $self );

		exists($$self{VERSION})
	and	$$self{VERSION} < $COMPATIBLE
	and	croak "Yapp driver version $VERSION ".
			  "incompatible with version $$self{VERSION}:\n".
			  "Please recompile parser module.";

        ref($class)
    and $class=ref($class);

    bless($self,$class);
}

sub YYParse {
    my($self)=shift;
    my($retval);

	_CheckParams( \@params, \%params, \@_, $self );

	if($$self{DEBUG}) {
		_DBLoad();
		$retval = eval '$self->_DBParse()';#Do not create stab entry on compile
        $@ and die $@;
	}
	else {
		$retval = $self->_Parse();
	}
    $retval
}

sub YYData {
	my($self)=shift;

		exists($$self{USER})
	or	$$self{USER}={};

	$$self{USER};
	
}

sub YYErrok {
	my($self)=shift;

	${$$self{ERRST}}=0;
    undef;
}

sub YYNberr {
	my($self)=shift;

	${$$self{NBERR}};
}

sub YYRecovering {
	my($self)=shift;

	${$$self{ERRST}} != 0;
}

sub YYAbort {
	my($self)=shift;

	${$$self{CHECK}}='ABORT';
    undef;
}

sub YYAccept {
	my($self)=shift;

	${$$self{CHECK}}='ACCEPT';
    undef;
}

sub YYError {
	my($self)=shift;

	${$$self{CHECK}}='ERROR';
    undef;
}

sub YYSemval {
	my($self)=shift;
	my($index)= $_[0] - ${$$self{DOTPOS}} - 1;

		$index < 0
	and	-$index <= @{$$self{STACK}}
	and	return $$self{STACK}[$index][1];

	undef;	#Invalid index
}

sub YYCurtok {
	my($self)=shift;

        @_
    and ${$$self{TOKEN}}=$_[0];
    ${$$self{TOKEN}};
}

sub YYCurval {
	my($self)=shift;

        @_
    and ${$$self{VALUE}}=$_[0];
    ${$$self{VALUE}};
}

sub YYExpect {
    my($self)=shift;

    keys %{$self->{STATES}[$self->{STACK}[-1][0]]{ACTIONS}}
}

sub YYLexer {
    my($self)=shift;

	$$self{LEX};
}


#################
# Private stuff #
#################


sub _CheckParams {
	my($mandatory,$checklist,$inarray,$outhash)=@_;
	my($prm,$value);
	my($prmlst)={};

	while(($prm,$value)=splice(@$inarray,0,2)) {
        $prm=uc($prm);
			exists($$checklist{$prm})
		or	croak("Unknow parameter '$prm'");
			ref($value) eq $$checklist{$prm}
		or	croak("Invalid value for parameter '$prm'");
        $prm=unpack('@2A*',$prm);
		$$outhash{$prm}=$value;
	}
	for (@$mandatory) {
			exists($$outhash{$_})
		or	croak("Missing mandatory parameter '".lc($_)."'");
	}
}

sub _Error {
	print "Parse error.\n";
}

sub _DBLoad {
	{
		no strict 'refs';

			exists(${__PACKAGE__.'::'}{_DBParse})#Already loaded ?
		and	return;
	}
	my($fname)=__FILE__;
	my(@drv);
	open(DRV,"<$fname") or die "Report this as a BUG: Cannot open $fname";
	while(<DRV>) {
                	/^\s*sub\s+_Parse\s*{\s*$/ .. /^\s*}\s*#\s*_Parse\s*$/
        	and     do {
                	s/^#DBG>//;
                	push(@drv,$_);
        	}
	}
	close(DRV);

	$drv[0]=~s/_P/_DBP/;
	eval join('',@drv);
}

#Note that for loading debugging version of the driver,
#this file will be parsed from 'sub _Parse' up to '}#_Parse' inclusive.
#So, DO NOT remove comment at end of sub !!!
sub _Parse {
    my($self)=shift;

	my($rules,$states,$lex,$error)
     = @$self{ 'RULES', 'STATES', 'LEX', 'ERROR' };
	my($errstatus,$nberror,$token,$value,$stack,$check,$dotpos)
     = @$self{ 'ERRST', 'NBERR', 'TOKEN', 'VALUE', 'STACK', 'CHECK', 'DOTPOS' };

#DBG>	my($debug)=$$self{DEBUG};
#DBG>	my($dbgerror)=0;

#DBG>	my($ShowCurToken) = sub {
#DBG>		my($tok)='>';
#DBG>		for (split('',$$token)) {
#DBG>			$tok.=		(ord($_) < 32 or ord($_) > 126)
#DBG>					?	sprintf('<%02X>',ord($_))
#DBG>					:	$_;
#DBG>		}
#DBG>		$tok.='<';
#DBG>	};

	$$errstatus=0;
	$$nberror=0;
	($$token,$$value)=(undef,undef);
	@$stack=( [ 0, undef ] );
	$$check='';

    while(1) {
        my($actions,$act,$stateno);

        $stateno=$$stack[-1][0];
        $actions=$$states[$stateno];

#DBG>	print STDERR ('-' x 40),"\n";
#DBG>		$debug & 0x2
#DBG>	and	print STDERR "In state $stateno:\n";
#DBG>		$debug & 0x08
#DBG>	and	print STDERR "Stack:[".
#DBG>					 join(',',map { $$_[0] } @$stack).
#DBG>					 "]\n";


        if  (exists($$actions{ACTIONS})) {

				defined($$token)
            or	do {
				($$token,$$value)=&$lex($self);
#DBG>				$debug & 0x01
#DBG>			and	print STDERR "Need token. Got ".&$ShowCurToken."\n";
			};

            $act=   exists($$actions{ACTIONS}{$$token})
                    ?   $$actions{ACTIONS}{$$token}
                    :   exists($$actions{DEFAULT})
                        ?   $$actions{DEFAULT}
                        :   undef;
        }
        else {
            $act=$$actions{DEFAULT};
#DBG>			$debug & 0x01
#DBG>		and	print STDERR "Don't need token.\n";
        }

            defined($act)
        and do {

                $act > 0
            and do {        #shift

#DBG>				$debug & 0x04
#DBG>			and	print STDERR "Shift and go to state $act.\n";

					$$errstatus
				and	do {
					--$$errstatus;

#DBG>					$debug & 0x10
#DBG>				and	$dbgerror
#DBG>				and	$$errstatus == 0
#DBG>				and	do {
#DBG>					print STDERR "**End of Error recovery.\n";
#DBG>					$dbgerror=0;
#DBG>				};
				};


                push(@$stack,[ $act, $$value ]);

					$$token ne ''	#Don't eat the eof
				and	$$token=$$value=undef;
                next;
            };

            #reduce
            my($lhs,$len,$code,@sempar,$semval);
            ($lhs,$len,$code)=@{$$rules[-$act]};

#DBG>			$debug & 0x04
#DBG>		and	$act
#DBG>		and	print STDERR "Reduce using rule ".-$act." ($lhs,$len): ";

                $act
            or  $self->YYAccept();

            $$dotpos=$len;

                unpack('A1',$lhs) eq '@'    #In line rule
            and do {
                    $lhs =~ /^\@[0-9]+\-([0-9]+)$/
                or  die "In line rule name '$lhs' ill formed: ".
                        "report it as a BUG.\n";
                $$dotpos = $1;
            };

            @sempar =       $$dotpos
                        ?   map { $$_[1] } @$stack[ -$$dotpos .. -1 ]
                        :   ();

            $semval = $code ? &$code( $self, @sempar )
                            : @sempar ? $sempar[0] : undef;

            splice(@$stack,-$len,$len);

                $$check eq 'ACCEPT'
            and do {

#DBG>			$debug & 0x04
#DBG>		and	print STDERR "Accept.\n";

				return($semval);
			};

                $$check eq 'ABORT'
            and	do {

#DBG>			$debug & 0x04
#DBG>		and	print STDERR "Abort.\n";

				return(undef);

			};

#DBG>			$debug & 0x04
#DBG>		and	print STDERR "Back to state $$stack[-1][0], then ";

                $$check eq 'ERROR'
            or  do {
#DBG>				$debug & 0x04
#DBG>			and	print STDERR 
#DBG>				    "go to state $$states[$$stack[-1][0]]{GOTOS}{$lhs}.\n";

#DBG>				$debug & 0x10
#DBG>			and	$dbgerror
#DBG>			and	$$errstatus == 0
#DBG>			and	do {
#DBG>				print STDERR "**End of Error recovery.\n";
#DBG>				$dbgerror=0;
#DBG>			};

			    push(@$stack,
                     [ $$states[$$stack[-1][0]]{GOTOS}{$lhs}, $semval ]);
                $$check='';
                next;
            };

#DBG>			$debug & 0x04
#DBG>		and	print STDERR "Forced Error recovery.\n";

            $$check='';

        };

        #Error
            $$errstatus
        or   do {

            $$errstatus = 1;
            &$error($self);
                $$errstatus # if 0, then YYErrok has been called
            or  next;       # so continue parsing

#DBG>			$debug & 0x10
#DBG>		and	do {
#DBG>			print STDERR "**Entering Error recovery.\n";
#DBG>			++$dbgerror;
#DBG>		};

            ++$$nberror;

        };

			$$errstatus == 3	#The next token is not valid: discard it
		and	do {
				$$token eq ''	# End of input: no hope
			and	do {
#DBG>				$debug & 0x10
#DBG>			and	print STDERR "**At eof: aborting.\n";
				return(undef);
			};

#DBG>			$debug & 0x10
#DBG>		and	print STDERR "**Dicard invalid token ".&$ShowCurToken.".\n";

			$$token=$$value=undef;
		};

        $$errstatus=3;

		while(	  @$stack
			  and (		not exists($$states[$$stack[-1][0]]{ACTIONS})
			        or  not exists($$states[$$stack[-1][0]]{ACTIONS}{error})
					or	$$states[$$stack[-1][0]]{ACTIONS}{error} <= 0)) {

#DBG>			$debug & 0x10
#DBG>		and	print STDERR "**Pop state $$stack[-1][0].\n";

			pop(@$stack);
		}

			@$stack
		or	do {

#DBG>			$debug & 0x10
#DBG>		and	print STDERR "**No state left on stack: aborting.\n";

			return(undef);
		};

		#shift the error token

#DBG>			$debug & 0x10
#DBG>		and	print STDERR "**Shift \$error token and go to state ".
#DBG>						 $$states[$$stack[-1][0]]{ACTIONS}{error}.
#DBG>						 ".\n";

		push(@$stack, [ $$states[$$stack[-1][0]]{ACTIONS}{error}, undef ]);

    }

    #never reached
	croak("Error in driver logic. Please, report it as a BUG");

}#_Parse
#DO NOT remove comment

1;

}
#End of include--------------------------------------------------




sub new {
        my($class)=shift;
        ref($class)
    and $class=ref($class);

    my($self)=$class->SUPER::new( yyversion => '1.21',
                                  yystates =>
[
	{#State 0
		ACTIONS => {
			'T_string' => 2,
			"<" => 4,
			"{" => 6
		},
		DEFAULT => -8,
		GOTOS => {
			'cell' => 1,
			'field' => 3,
			'portSpec' => 5,
			'cellList' => 7
		}
	},
	{#State 1
		DEFAULT => -3
	},
	{#State 2
		DEFAULT => -7
	},
	{#State 3
		ACTIONS => {
			'' => 8
		}
	},
	{#State 4
		ACTIONS => {
			'T_string' => 9
		}
	},
	{#State 5
		ACTIONS => {
			'T_string' => 10
		},
		DEFAULT => -6
	},
	{#State 6
		ACTIONS => {
			'T_string' => 2,
			"<" => 4
		},
		DEFAULT => -8,
		GOTOS => {
			'cell' => 1,
			'portSpec' => 5,
			'cellList' => 11
		}
	},
	{#State 7
		ACTIONS => {
			"|" => 12
		},
		DEFAULT => -1
	},
	{#State 8
		DEFAULT => 0
	},
	{#State 9
		ACTIONS => {
			">" => 13
		}
	},
	{#State 10
		DEFAULT => -5
	},
	{#State 11
		ACTIONS => {
			"}" => 14,
			"|" => 12
		}
	},
	{#State 12
		ACTIONS => {
			'T_string' => 2,
			"<" => 4
		},
		DEFAULT => -8,
		GOTOS => {
			'cell' => 15,
			'portSpec' => 5
		}
	},
	{#State 13
		DEFAULT => -9
	},
	{#State 14
		DEFAULT => -2
	},
	{#State 15
		DEFAULT => -4
	}
],
                                  yyrules  =>
[
	[#Rule 0
		 '$start', 2, undef
	],
	[#Rule 1
		 'field', 1,
sub
#line 11 "parseRecordLabel.yp"
{ $_[1] }
	],
	[#Rule 2
		 'field', 3,
sub
#line 12 "parseRecordLabel.yp"
{ $_[2] }
	],
	[#Rule 3
		 'cellList', 1,
sub
#line 15 "parseRecordLabel.yp"
{ [ $_[1] ] }
	],
	[#Rule 4
		 'cellList', 3,
sub
#line 16 "parseRecordLabel.yp"
{ [ @{ $_[1] }, $_[3] ] }
	],
	[#Rule 5
		 'cell', 2,
sub
#line 19 "parseRecordLabel.yp"
{ +{ $_[1], $_[2] } }
	],
	[#Rule 6
		 'cell', 1,
sub
#line 20 "parseRecordLabel.yp"
{ +{ $_[1], '' } }
	],
	[#Rule 7
		 'cell', 1,
sub
#line 21 "parseRecordLabel.yp"
{ +{ '', $_[1] } }
	],
	[#Rule 8
		 'cell', 0,
sub
#line 22 "parseRecordLabel.yp"
{ +{ '', '' } }
	],
	[#Rule 9
		 'portSpec', 3,
sub
#line 25 "parseRecordLabel.yp"
{ $_[2] }
	]
],
                                  @_);
    bless($self,$class);
}

#line 28 "parseRecordLabel.yp"


sub Error {
    my $parse = shift;
    my($token)=$parse->YYCurtok;
    my($value)=$parse->YYCurval;
    my($expected)=$parse->YYExpect;

    my $input = $parse->YYData->{INPUT};

    # Get rid of all but the first line
    ($input) = split("\n",$input);

    die <<EOF;
Parse Error, Got token/value '$token', '$value'; Expected token '$expected'
Near line :
$input
EOF
}

sub Lexer {
    my($parser)=shift;

    my @expect = $parser->YYExpect;

    # If at the end of the string, and expecting a T_string token
    #   Return a null t_string
    #   This enables strings like '<f0> 0x10ba8| <f1>' to be parsed
    #   correctly
    if( $parser->YYData->{INPUT} eq '' && @expect == 1 && $expect[0] eq 'T_string'){
	return('T_string','');
    }

    defined($parser->YYData->{INPUT})
    or  return('',undef);

    for( $parser->YYData->{INPUT}){
	# Different Token Types

	# check for tokens '<>{} tokens (Whitespace OK)
	if( s/^\s*([\<\>\{])\s*//){  # <, > and { with whitespace around
		return($1, $1);
	}
	if( s/^(\})\s*//){  # }  with whitespace after
		return($1, $1);
	}
	if( s/^(\|)//){  # |  with no whitespace
		return($1, $1);
	}

	# T_string
	s/^(.*?)((?<!\\)[\>\{\|\}])/$2/s  # strings with embedded special characters (not backslashed)
	and return('T_string',$1);

	# End of string, return everything
	s/(.+)//s
		and return ('T_string', $1);

	return('','');
  }
}

1;
